devito/function.py

```diff
@@ -330,50 +284,67 @@ def argument_values(self, alias=None, **kwargs):
         # Add value override for own data if it is provided
         if self.name in kwargs:
             new = kwargs.pop(self.name)
-            if len(new.shape) != self.ndim:
-                raise ValueError("Array shape %s does not match" % (new.shape, ) +
-                                 "dimensions %s" % (self.indices, ))
             if isinstance(new, TensorFunction):
                 # Set new values and re-derive defaults
                 values[key.name] = new._data_buffer
-                values.update(new.argument_defaults(alias=key).reduce_all())
+                values.update(new._arg_defaults(alias=key).reduce_all())
             else:
                 # We've been provided a pure-data replacement (array)
                 values[key.name] = new
                 # Add value overrides for all associated dimensions
                 for i, s, o in zip(self.indices, new.shape, self.staggered):
-                    values.update(i.argument_defaults(size=s+o))
+                    values.update(i._arg_defaults(size=s+o-sum(self._offset_domain[i])))
```

Comment:
Why is `self._offset_domain[i]` necessary here but not in `_args_defaults`?

Reply:
ah, good question ! 
so this ``else`` branch handles the case in which a numpy array is passed in. When we generate the dimension values, however, we should do that w.r.t. the actual domain, not w.r.t. the domain+halo+padding; hence, with ``-sum(...)`` we cut away the region between the last point of the domain and the end of the whole array

Reply:
IOW, in ``_args_default`` you don't need it because for ``self`` you know its shape (which, I recall, is generally smaller than ``shape_with_halo``, which in turn is generally smaller than ``shape_allocated`` -- the latter being the actual shape of the whole array)